// sane/src/watchman_watcher.js
'use strict';

var fs = require('fs');
var path = require('path');
var common = require('./common');
var watchmanClient = require('./watchman_client');
var EventEmitter = require('events').EventEmitter;
var RecrawlWarning = require('./utils/recrawl-warning-dedupe');



var CHANGE_EVENT = common.CHANGE_EVENT;
var DELETE_EVENT = common.DELETE_EVENT;
var ADD_EVENT = common.ADD_EVENT;
var ALL_EVENT = common.ALL_EVENT;



module.exports = WatchmanWatcher;



function WatchmanWatcher(dir, opts) {
  common.assignOptions(this, opts);
  this.root = path.resolve(dir);
  this._init();
}

WatchmanWatcher.prototype.__proto__ = EventEmitter.prototype;


WatchmanWatcher.prototype._init = function() {
  if (this._client) {
    this._client = null;
  }

  // Get the WatchmanClient instance corresponding to our watchmanPath (or nothing).
  // Then subscribe, which will do the appropriate setup so that we will receive
  // calls to handleChangeEvent when files change.
  this._client = watchmanClient.getInstance(this.watchmanPath);

  return this._client.subscribe(this, this.root).then(
    resp => {
      this._handleWarning(resp);
      this.emit('ready');
    },
    error => {
      this._handleError(error);
    }
  );
};


WatchmanWatcher.prototype.createOptions = function() {
  let options = {
    fields: ['name', 'exists', 'new'],
  };

  // If the server has the wildmatch capability available it supports
  // the recursive **
WatchmanWatcher.prototype.handleErrorEvent = function(error) {
  this.emit('error', error);
};



WatchmanWatcher.prototype.handleChangeEvent = function(resp) {
  if (Array.isArray(resp.files)) {
    resp.files.forEach(this.handleFileChange, this);
  }
};



WatchmanWatcher.prototype.handleFileChange = function(changeDescriptor) {
  var absPath;
  var relativePath;

  relativePath = changeDescriptor.name;
  absPath = path.join(this.root, relativePath);

  if (
    !(this._client.wildmatch && !this.hasIgnore) &&
    !common.isFileIncluded(this.globs, this.dot, this.doIgnore, relativePath)
  ) {
    return;
  }

  if (!changeDescriptor.exists) {
    this.emitEvent(DELETE_EVENT, relativePath, this.root);
  } else {
    fs.lstat(absPath, (error, stat) => {
      // Files can be deleted between the event and the lstat call
      // the most reliable thing to do here is to ignore the event.
      if (error && error.code === 'ENOENT') {
        return;
      }

      if (this._handleError(error)) {
        return;
      }

      var eventType = changeDescriptor.new ? ADD_EVENT : CHANGE_EVENT;

      // Change event on dirs are mostly useless.
      if (!(eventType === CHANGE_EVENT && stat.isDirectory())) {
        this.emitEvent(eventType, relativePath, this.root, stat);
      }
    });
  }
};



WatchmanWatcher.prototype.emitEvent = function(
  eventType,
  filepath,
  root,
  stat
) {
  this.emit(eventType, filepath, root, stat);
  this.emit(ALL_EVENT, eventType, filepath, root, stat);
};



WatchmanWatcher.prototype.close = function(callback) {
  this._client.closeWatcher(this);
  callback && callback(null, true);
};



WatchmanWatcher.prototype._handleError = function(error) {
  if (error != null) {
    this.emit('error', error);
    return true;
  } else {
    return false;
  }
};



WatchmanWatcher.prototype._handleWarning = function(resp) {
  if ('warning' in resp) {
    if (RecrawlWarning.isRecrawlWarningDupe(resp.warning)) {
      return true;
    }
    console.warn(resp.warning);
    return true;
  } else {
    return false;
  }
};
