// sane/src/watchman_client.js
'use strict';

var watchman = require('fb-watchman');
var captureExit = require('capture-exit');

function values(obj) {
  return Object.keys(obj).map(key => obj[key]);
}





function WatchmanClient(watchmanBinaryPath) {
  captureExit.captureExit();

  // define/clear some local state. The properties will be initialized
  // in _handleClientAndCheck(). This is also called again in _onEnd when
  // trying to reestablish connection to watchman.
  this._clearLocalVars();

  this._watchmanBinaryPath = watchmanBinaryPath;

  this._backoffTimes = this._setupBackoffTimes();

  this._clientListeners = null; // direct listeners from here to watchman.Client.

  // Define a handler for if somehow the Node process gets interrupted. We need to
  // close down the watchman.Client, if we have one.
  captureExit.onExit(() => {
    this._clearLocalVars();
  });
}

// Define 'wildmatch' property, which must be available when we call the
// WatchmanWatcher.createOptions() method.
Object.defineProperty(WatchmanClient.prototype, 'wildmatch', {
  get() {
    return this._wildmatch;
  },
});


WatchmanClient.prototype.subscribe = function(watchmanWatcher, root) {
  let subscription;
  let watcherInfo;

  return this._setupClient()
    .then(() => {
      watcherInfo = this._createWatcherInfo(watchmanWatcher);
      subscription = watcherInfo.subscription;
      return this._watch(subscription, root);
    })
    .then(() => this._clock(subscription))
    .then(() => this._subscribe(subscription));
  // Note: callers are responsible for noting any subscription failure.
};


WatchmanClient.prototype.closeWatcher = function(watchmanWatcher) {
  let watcherInfos = values(this._watcherMap);
  let numWatchers = watcherInfos.length;

  if (numWatchers > 0) {
    let watcherInfo;

    for (let info of watcherInfos) {
      if (info.watchmanWatcher === watchmanWatcher) {
        watcherInfo = info;
        break;
      }
    }

    if (watcherInfo) {
      delete this._watcherMap[watcherInfo.subscription];

      numWatchers--;

      if (numWatchers === 0) {
        this._clearLocalVars(); // nobody watching, so shut the watchman.Client down.
      }
    }
  }
};


WatchmanClient.prototype._setupBackoffTimes = function() {
  return {
    _times: [0, 1000, 5000, 10000, 60000],

    _next: 0,

    next() {
      let val = this._times[this._next];
      if (this._next < this._times.length - 1) {
        this._next++;
      }
      return val;
    },

    reset() {
      this._next = 0;
    },
  };
};


WatchmanClient.prototype._setupClient = function() {
  if (!this._clientPromise) {
    this._clientPromise = new Promise((resolve, reject) => {
      this._handleClientAndCheck(resolve, reject);
    });
  }

  return this._clientPromise;
};


WatchmanClient.prototype._handleClientAndCheck = function(resolve, reject) {
  this._createClientAndCheck().then(
    value => {
      let resp = value.resp;
      let client = value.client;

      try {
        this._wildmatch = resp.capabilities.wildmatch;
        this._relative_root = resp.capabilities.relative_root;
        this._client = client;

        client.on('subscription', this._onSubscription.bind(this));
        client.on('error', this._onError.bind(this));
        client.on('end', this._onEnd.bind(this));

        this._backoffTimes.reset();
        resolve(this);
      } catch (error) {
        // somehow, even though we supposedly got a valid value back, it's
        // malformed, or some other internal error occurred. Reject so
        // the promise itself doesn't hang forever.
        reject(error);
      }
    },
    () => {
      // create & capability check failed in any of several ways,
      // do the retry with backoff.

      // XXX May want to change this later to actually reject/terminate with
      // an error in certain of the inner errors (e.g. when we
      // can figure out the server is definitely not coming
      // back, or something else is not recoverable by just waiting).
      // Could also decide after N retries to just quit.

      let backoffMillis = this._backoffTimes.next();

      // XXX may want to fact we'll attempt reconnect in backoffMillis ms.
      setTimeout(() => {
        this._handleClientAndCheck(resolve, reject);
      }, backoffMillis);
    }
  );
};


WatchmanClient.prototype._createClientAndCheck = function() {
  return new Promise((resolve, reject) => {
    let client;

    try {
      client = new watchman.Client(
        this._watchmanBinaryPath
          ? { watchmanBinaryPath: this._watchmanBinaryPath }
          : {}
      );
    } catch (error) {
      // if we're here, either the binary path is bad or something
      // else really bad happened. The client doesn't even attempt
      // to connect until we send it a command, though.
      reject(error);
      return;
    }

    client.on('error', error => {
      client.removeAllListeners();
      reject(error);
    });

    client.on('end', () => {
      client.removeAllListeners();
      reject(new Error('Disconnected during client capabilities check'));
    });

    client.capabilityCheck(
      { optional: ['wildmatch', 'relative_root'] },
      (error, resp) => {
        try {
          client.removeAllListeners();

          if (error) {
            reject(error);
          } else {
            resolve({ resp, client });
          }
        } catch (err) {
          // In case we get something weird in the block using 'resp'.
          // XXX We could also just remove the try/catch if we believe
          // the resp stuff should always work, but just in case...
          reject(err);
        }
      }
    );
  });
};


WatchmanClient.prototype._clearLocalVars = function() {
  if (this._client) {
    this._client.removeAllListeners();
    this._client.end();
  }

  this._client = null;
  this._clientPromise = null;
  this._wildmatch = false;
  this._relative_root = false;
  this._subscriptionId = 1;
  this._watcherMap = Object.create(null);

  // Note that we do not clear _clientListeners or _watchmanBinaryPath.
};

WatchmanClient.prototype._genSubscription = function() {
  let val = 'sane_' + this._subscriptionId++;
  return val;
};


WatchmanClient.prototype._createWatcherInfo = function(watchmanWatcher) {
  let watcherInfo = {
    subscription: this._genSubscription(),
    watchmanWatcher: watchmanWatcher,
    root: null, // set during 'watch' or 'watch-project'
    relativePath: null, // same
    since: null, // set during 'clock'
    options: null, // created and set during 'subscribe'.
  };

  this._watcherMap[watcherInfo.subscription] = watcherInfo;

  return watcherInfo;
};


WatchmanClient.prototype._getWatcherInfo = function(subscription) {
  return this._watcherMap[subscription];
};


WatchmanClient.prototype._watch = function(subscription, root) {
  return new Promise((resolve, reject) => {
    let watcherInfo = this._getWatcherInfo(subscription);

    if (this._relative_root) {
      this._client.command(['watch-project', root], (error, resp) => {
        if (error) {
          reject(error);
        } else {
          watcherInfo.root = resp.watch;
          watcherInfo.relativePath = resp.relative_path
            ? resp.relative_path
            : '';
          resolve(resp);
        }
      });
    } else {
      this._client.command(['watch', root], (error, resp) => {
        if (error) {
          reject(error);
        } else {
          watcherInfo.root = root;
          watcherInfo.relativePath = '';
          resolve(resp);
        }
      });
    }
  });
};


WatchmanClient.prototype._clock = function(subscription) {
  return new Promise((resolve, reject) => {
    let watcherInfo = this._getWatcherInfo(subscription);

    this._client.command(['clock', watcherInfo.root], (error, resp) => {
      if (error) {
        reject(error);
      } else {
        watcherInfo.since = resp.clock;
        resolve(resp);
      }
    });
  });
};


WatchmanClient.prototype._subscribe = function(subscription) {
  return new Promise((resolve, reject) => {
    let watcherInfo = this._getWatcherInfo(subscription);

    // create the 'bare' options w/o 'since' or relative_root.
    // Store in watcherInfo for later use if we need to reset
    // things after an 'end' caught here.
    let options = watcherInfo.watchmanWatcher.createOptions();
    watcherInfo.options = options;

    // Dup the options object so we can add 'relative_root' and 'since'
    // and leave the original options object alone. We'll do this again
    // later if we need to resubscribe after 'end' and reconnect.
    options = Object.assign({}, options);

    if (this._relative_root) {
      options.relative_root = watcherInfo.relativePath;
    }

    options.since = watcherInfo.since;

    this._client.command(
      ['subscribe', watcherInfo.root, subscription, options],
      (error, resp) => {
        if (error) {
          reject(error);
        } else {
          resolve(resp);
        }
      }
    );
  });
};


WatchmanClient.prototype._onSubscription = function(resp) {
  let watcherInfo = this._getWatcherInfo(resp.subscription);

  if (watcherInfo) {
    // we're assuming the watchmanWatcher does not throw during
    // handling of the change event.
    watcherInfo.watchmanWatcher.handleChangeEvent(resp);
  } else {
    // Note it in the log, but otherwise ignore it
    console.error(
      "WatchmanClient error - received 'subscription' event " +
        "for non-existent subscription '" +
        resp.subscription +
        "'"
    );
  }
};


WatchmanClient.prototype._onError = function(error) {
  values(this._watcherMap).forEach(watcherInfo =>
    watcherInfo.watchmanWatcher.handleErrorEvent(error)
  );
};


WatchmanClient.prototype._onEnd = function() {
  console.warn(
    '[sane.WatchmanClient] Warning: Lost connection to watchman, reconnecting..'
  );

  // Hold the old watcher map so we use it to recreate all subscriptions.
  let oldWatcherInfos = values(this._watcherMap);

  this._clearLocalVars();

  this._setupClient().then(
    () => {
      let promises = oldWatcherInfos.map(watcherInfo =>
        this.subscribe(
          watcherInfo.watchmanWatcher,
          watcherInfo.watchmanWatcher.root
        )
      );
      Promise.all(promises).then(
        () => {
          console.log('[sane.WatchmanClient]: Reconnected to watchman');
        },
        error => {
          console.error(
            '[sane.WatchmanClient]: Reconnected to watchman, but failed to ' +
              'reestablish at least one subscription, cannot continue'
          );
          console.error(error);
          oldWatcherInfos.forEach(watcherInfo =>
            watcherInfo.watchmanWatcher.handleErrorEvent(error)
          );
          // XXX not sure whether to clear all _watcherMap instances here,
          // but basically this client is inconsistent now, since at least one
          // subscribe failed.
        }
      );
    },
    error => {
      console.error(
        '[sane.WatchmanClient]: Lost connection to watchman, ' +
          'reconnect failed, cannot continue'
      );
      console.error(error);
      oldWatcherInfos.forEach(watcherInfo =>
        watcherInfo.watchmanWatcher.handleErrorEvent(error)
      );
    }
  );
};

module.exports = {
  
  getInstance: function(watchmanBinaryPath) {
    let clientMap = WatchmanClient.prototype._clientMap;

    if (!clientMap) {
      clientMap = Object.create(null);
      WatchmanClient.prototype._clientMap = clientMap;
    }

    if (watchmanBinaryPath == undefined || watchmanBinaryPath === null) {
      watchmanBinaryPath = '';
    }

    let watchmanClient = clientMap[watchmanBinaryPath];

    if (!watchmanClient) {
      watchmanClient = new WatchmanClient(watchmanBinaryPath);
      clientMap[watchmanBinaryPath] = watchmanClient;
    }

    return watchmanClient;
  },
};
