// jsdom/lib/jsdom/living/constraint-validation/DefaultConstraintValidation-impl.js
"use strict";

const ValidityState = require("../generated/ValidityState");
const Event = require("../generated/Event");
const { isDisabled } = require("../helpers/form-controls");
const { closest } = require("../helpers/traversal");

exports.implementation = class DefaultConstraintValidationImpl {
  
  get willValidate() {
    return this._isCandidateForConstraintValidation();
  }

  get validity() {
    if (!this._validity) {
      this._validity = ValidityState.createImpl(this);
    }
    return this._validity;
  }

  
  checkValidity() {
    if (!this._isCandidateForConstraintValidation()) {
      return true;
    }
    if (this._satisfiesConstraints()) {
      return true;
    }
    this.dispatchEvent(Event.createImpl(["invalid", { cancelable: true }]));
    return false;
  }

  
  setCustomValidity(message) {
    this._customValidityErrorMessage = message;
  }

  
  
  reportValidity() {
    return this.checkValidity();
  }

  
  get validationMessage() {
    const { validity } = this;
    if (!this._isCandidateForConstraintValidation() || this._satisfiesConstraints()) {
      return "";
    }
    const isSufferingFromCustomError = validity.customError;
    if (isSufferingFromCustomError) {
      return this._customValidityErrorMessage;
    }
    return "Constraints not satisfied";
  }

  _isCandidateForConstraintValidation() {
    
    return !isDisabled(this) &&
      
      
      closest(this, "datalist") === null &&
      !this._barredFromConstraintValidationSpecialization();
  }

  _isBarredFromConstraintValidation() {
    return !this._isCandidateForConstraintValidation();
  }

  _satisfiesConstraints() {
    return this.validity.valid;
  }
};
