// jsdom/lib/jsdom/living/helpers/ordered-set.js
"use strict";






module.exports = class OrderedSet {
  constructor() {
    this._items = [];
  }

  append(item) {
    if (!this.contains(item)) {
      this._items.push(item);
    }
  }

  prepend(item) {
    if (!this.contains(item)) {
      this._items.unshift(item);
    }
  }

  replace(item, replacement) {
    let seen = false;
    for (let i = 0; i < this._items.length;) {
      const isInstance = this._items[i] === item || this._items[i] === replacement;
      if (seen && isInstance) {
        this._items.splice(i, 1);
      } else {
        if (isInstance) {
          this._items[i] = replacement;
          seen = true;
        }
        i++;
      }
    }
  }

  remove(...items) {
    this.removePredicate(item => items.includes(item));
  }

  removePredicate(predicate) {
    for (let i = 0; i < this._items.length;) {
      if (predicate(this._items[i])) {
        this._items.splice(i, 1);
      } else {
        i++;
      }
    }
  }

  empty() {
    this._items.length = 0;
  }

  contains(item) {
    return this._items.includes(item);
  }

  get size() {
    return this._items.length;
  }

  isEmpty() {
    return this._items.length === 0;
  }

  

  [Symbol.iterator]() {
    return this._items[Symbol.iterator]();
  }

  keys() {
    return this._items.keys();
  }

  get(index) {
    return this._items[index];
  }

  some(func) {
    return this._items.some(func);
  }

  
  static parse(input) {
    const tokens = new OrderedSet();
    for (const token of input.split(/[\t\n\f\r ]+/)) {
      if (token) {
        tokens.append(token);
      }
    }
    return tokens;
  }

  
  serialize() {
    return this._items.join(" ");
  }
};
