// istanbul-api/index.js

var Reporter = require('./lib/reporter');

module.exports = {
    config: require('./lib/config'),
    cover: require('./lib/run-cover'),
    reports: require('./lib/run-reports'),
    instrument: require('./lib/run-instrument'),
    checkCoverage: require('./lib/run-check-coverage'),
    createReporter: function (cfg, opts) {
        return new Reporter(cfg, opts);
    },
    
    matcherFor: require('./lib/file-matcher').matcherFor,
    filesFor: require('./lib/file-matcher').filesFor
};

// export all the istanbul libraries as is so users don't have to take 5 deps
// that are potentially inconsistent

var DASH_PATTERN = /-([a-z])/g;

function camelize(word) {
    return word.replace(DASH_PATTERN, function (match, lch) {
        return lch.toUpperCase();
    });
}

[ 'coverage', 'hook', 'instrument', 'report', 'source-maps'].forEach(function (k) {
        var mod = 'lib-' + k,
            prop = camelize(mod);
        module.exports[prop] = require('istanbul-' + mod);
});

module.exports.reportsImpl = require('istanbul-reports');
