// es-abstract/operations/2015.js
'use strict';

module.exports = {
	IsPropertyDescriptor: 'https:

	abs: {
		url: 'https:
	},
	'Abstract Equality Comparison': {
		url: 'https:
	},
	'Abstract Relational Comparison': {
		url: 'https:
	},
	AddRestrictedFunctionProperties: {
		url: 'https:
	},
	AdvanceStringIndex: {
		url: 'https:
	},
	AllocateArrayBuffer: {
		url: 'https:
	},
	AllocateTypedArray: {
		url: 'https:
	},
	ArrayCreate: {
		url: 'https:
	},
	ArraySetLength: {
		url: 'https:
	},
	ArraySpeciesCreate: {
		url: 'https:
	},
	BoundFunctionCreate: {
		url: 'https:
	},
	Call: {
		url: 'https:
	},
	Canonicalize: {
		url: 'https:
	},
	CanonicalNumericIndexString: {
		url: 'https:
	},
	CharacterRange: {
		url: 'https:
	},
	CharacterSetMatcher: {
		url: 'https:
	},
	CloneArrayBuffer: {
		url: 'https:
	},
	CompletePropertyDescriptor: {
		url: 'https:
	},
	Completion: {
		url: 'https:
	},
	CompletionRecord: {
		url: 'https:
	},
	Construct: {
		url: 'https:
	},
	CopyDataBlockBytes: {
		url: 'https:
	},
	CreateArrayFromList: {
		url: 'https:
	},
	CreateArrayIterator: {
		url: 'https:
	},
	CreateBuiltinFunction: {
		url: 'https:
	},
	CreateByteDataBlock: {
		url: 'https:
	},
	CreateDataProperty: {
		url: 'https:
	},
	CreateDataPropertyOrThrow: {
		url: 'https:
	},
	CreateDynamicFunction: {
		url: 'https:
	},
	CreateHTML: {
		url: 'https:
	},
	CreateIntrinsics: {
		url: 'https:
	},
	CreateIterResultObject: {
		url: 'https:
	},
	CreateListFromArrayLike: {
		url: 'https:
	},
	CreateListIterator: {
		url: 'https:
	},
	CreateMapIterator: {
		url: 'https:
	},
	CreateMappedArgumentsObject: {
		url: 'https:
	},
	CreateMethodProperty: {
		url: 'https:
	},
	CreatePerIterationEnvironment: {
		url: 'https:
	},
	CreateRealm: {
		url: 'https:
	},
	CreateSetIterator: {
		url: 'https:
	},
	CreateUnmappedArgumentsObject: {
		url: 'https:
	},
	DateFromTime: {
		url: 'https:
	},
	Day: {
		url: 'https:
	},
	DayFromYear: {
		url: 'https:
	},
	DaylightSavingTA: {
		url: 'https:
	},
	DaysInYear: {
		url: 'https:
	},
	DayWithinYear: {
		url: 'https:
	},
	Decode: {
		url: 'https:
	},
	DefinePropertyOrThrow: {
		url: 'https:
	},
	DeletePropertyOrThrow: {
		url: 'https:
	},
	DetachArrayBuffer: {
		url: 'https:
	},
	Encode: {
		url: 'https:
	},
	EnqueueJob: {
		url: 'https:
	},
	EnumerableOwnNames: {
		url: 'https:
	},
	EscapeRegExpPattern: {
		url: 'https:
	},
	EvalDeclarationInstantiation: {
		url: 'https:
	},
	EvaluateCall: {
		url: 'https:
	},
	EvaluateDirectCall: {
		url: 'https:
	},
	EvaluateNew: {
		url: 'https:
	},
	floor: {
		url: 'https:
	},
	ForBodyEvaluation: {
		url: 'https:
	},
	'ForIn/OfBodyEvaluation': {
		url: 'https:
	},
	'ForIn/OfHeadEvaluation': {
		url: 'https:
	},
	FromPropertyDescriptor: {
		url: 'https:
	},
	FulfillPromise: {
		url: 'https:
	},
	FunctionAllocate: {
		url: 'https:
	},
	FunctionCreate: {
		url: 'https:
	},
	FunctionInitialize: {
		url: 'https:
	},
	GeneratorFunctionCreate: {
		url: 'https:
	},
	GeneratorResume: {
		url: 'https:
	},
	GeneratorResumeAbrupt: {
		url: 'https:
	},
	GeneratorStart: {
		url: 'https:
	},
	GeneratorValidate: {
		url: 'https:
	},
	GeneratorYield: {
		url: 'https:
	},
	Get: {
		url: 'https:
	},
	GetBase: {
		url: 'https:
	},
	GetFunctionRealm: {
		url: 'https:
	},
	GetGlobalObject: {
		url: 'https:
	},
	GetIdentifierReference: {
		url: 'https:
	},
	GetIterator: {
		url: 'https:
	},
	GetMethod: {
		url: 'https:
	},
	GetModuleNamespace: {
		url: 'https:
	},
	GetNewTarget: {
		url: 'https:
	},
	GetOwnPropertyKeys: {
		url: 'https:
	},
	GetPrototypeFromConstructor: {
		url: 'https:
	},
	GetReferencedName: {
		url: 'https:
	},
	GetSubstitution: {
		url: 'https:
	},
	GetSuperConstructor: {
		url: 'https:
	},
	GetTemplateObject: {
		url: 'https:
	},
	GetThisEnvironment: {
		url: 'https:
	},
	GetThisValue: {
		url: 'https:
	},
	GetV: {
		url: 'https:
	},
	GetValue: {
		url: 'https:
	},
	GetValueFromBuffer: {
		url: 'https:
	},
	GetViewValue: {
		url: 'https:
	},
	HasOwnProperty: {
		url: 'https:
	},
	HasPrimitiveBase: {
		url: 'https:
	},
	HasProperty: {
		url: 'https:
	},
	HourFromTime: {
		url: 'https:
	},
	ImportedLocalNames: {
		url: 'https:
	},
	InitializeHostDefinedRealm: {
		url: 'https:
	},
	InitializeReferencedBinding: {
		url: 'https:
	},
	InLeapYear: {
		url: 'https:
	},
	InstanceofOperator: {
		url: 'https:
	},
	IntegerIndexedElementGet: {
		url: 'https:
	},
	IntegerIndexedElementSet: {
		url: 'https:
	},
	IntegerIndexedObjectCreate: {
		url: 'https:
	},
	InternalizeJSONProperty: {
		url: 'https:
	},
	Invoke: {
		url: 'https:
	},
	IsAccessorDescriptor: {
		url: 'https:
	},
	IsAnonymousFunctionDefinition: {
		url: 'https:
	},
	IsArray: {
		url: 'https:
	},
	IsCallable: {
		url: 'https:
	},
	IsCompatiblePropertyDescriptor: {
		url: 'https:
	},
	IsConcatSpreadable: {
		url: 'https:
	},
	IsConstructor: {
		url: 'https:
	},
	IsDataDescriptor: {
		url: 'https:
	},
	IsDetachedBuffer: {
		url: 'https:
	},
	IsExtensible: {
		url: 'https:
	},
	IsGenericDescriptor: {
		url: 'https:
	},
	IsInTailPosition: {
		url: 'https:
	},
	IsInteger: {
		url: 'https:
	},
	IsLabelledFunction: {
		url: 'https:
	},
	IsPromise: {
		url: 'https:
	},
	IsPropertyKey: {
		url: 'https:
	},
	IsPropertyReference: {
		url: 'https:
	},
	IsRegExp: {
		url: 'https:
	},
	IsStrictReference: {
		url: 'https:
	},
	IsSuperReference: {
		url: 'https:
	},
	IsUnresolvableReference: {
		url: 'https:
	},
	IsWordChar: {
		url: 'https:
	},
	IteratorClose: {
		url: 'https:
	},
	IteratorComplete: {
		url: 'https:
	},
	IteratorNext: {
		url: 'https:
	},
	IteratorStep: {
		url: 'https:
	},
	IteratorValue: {
		url: 'https:
	},
	LocalTime: {
		url: 'https:
	},
	LoopContinues: {
		url: 'https:
	},
	MakeArgGetter: {
		url: 'https:
	},
	MakeArgSetter: {
		url: 'https:
	},
	MakeClassConstructor: {
		url: 'https:
	},
	MakeConstructor: {
		url: 'https:
	},
	MakeDate: {
		url: 'https:
	},
	MakeDay: {
		url: 'https:
	},
	MakeMethod: {
		url: 'https:
	},
	MakeSuperPropertyReference: {
		url: 'https:
	},
	MakeTime: {
		url: 'https:
	},
	max: {
		url: 'https:
	},
	min: {
		url: 'https:
	},
	MinFromTime: {
		url: 'https:
	},
	ModuleNamespaceCreate: {
		url: 'https:
	},
	modulo: {
		url: 'https:
	},
	MonthFromTime: {
		url: 'https:
	},
	msFromTime: {
		url: 'https:
	},
	msPerDay: {
		url: 'https:
	},
	NewDeclarativeEnvironment: {
		url: 'https:
	},
	NewFunctionEnvironment: {
		url: 'https:
	},
	NewGlobalEnvironment: {
		url: 'https:
	},
	NewModuleEnvironment: {
		url: 'https:
	},
	NewObjectEnvironment: {
		url: 'https:
	},
	NewPromiseCapability: {
		url: 'https:
	},
	NormalCompletion: {
		url: 'https:
	},
	ObjectCreate: {
		url: 'https:
	},
	ObjectDefineProperties: {
		url: 'https:
	},
	OrdinaryCallBindThis: {
		url: 'https:
	},
	OrdinaryCallEvaluateBody: {
		url: 'https:
	},
	OrdinaryCreateFromConstructor: {
		url: 'https:
	},
	OrdinaryDefineOwnProperty: {
		url: 'https:
	},
	OrdinaryGetOwnProperty: {
		url: 'https:
	},
	OrdinaryHasInstance: {
		url: 'https:
	},
	OrdinaryHasProperty: {
		url: 'https:
	},
	ParseModule: {
		url: 'https:
	},
	PerformEval: {
		url: 'https:
	},
	PerformPromiseAll: {
		url: 'https:
	},
	PerformPromiseRace: {
		url: 'https:
	},
	PerformPromiseThen: {
		url: 'https:
	},
	PrepareForOrdinaryCall: {
		url: 'https:
	},
	PrepareForTailCall: {
		url: 'https:
	},
	ProxyCreate: {
		url: 'https:
	},
	PutValue: {
		url: 'https:
	},
	QuoteJSONString: {
		url: 'https:
	},
	RegExpAlloc: {
		url: 'https:
	},
	RegExpBuiltinExec: {
		url: 'https:
	},
	RegExpCreate: {
		url: 'https:
	},
	RegExpExec: {
		url: 'https:
	},
	RegExpInitialize: {
		url: 'https:
	},
	RejectPromise: {
		url: 'https:
	},
	RepeatMatcher: {
		url: 'https:
	},
	RequireObjectCoercible: {
		url: 'https:
	},
	ResolveBinding: {
		url: 'https:
	},
	ResolveThisBinding: {
		url: 'https:
	},
	SameValue: {
		url: 'https:
	},
	SameValueZero: {
		url: 'https:
	},
	SecFromTime: {
		url: 'https:
	},
	SerializeJSONArray: {
		url: 'https:
	},
	SerializeJSONObject: {
		url: 'https:
	},
	SerializeJSONProperty: {
		url: 'https:
	},
	Set: {
		url: 'https:
	},
	SetDefaultGlobalBindings: {
		url: 'https:
	},
	SetFunctionName: {
		url: 'https:
	},
	SetIntegrityLevel: {
		url: 'https:
	},
	SetRealmGlobalObject: {
		url: 'https:
	},
	SetValueInBuffer: {
		url: 'https:
	},
	SetViewValue: {
		url: 'https:
	},
	sign: {
		url: 'https:
	},
	SortCompare: {
		url: 'https:
	},
	SpeciesConstructor: {
		url: 'https:
	},
	SplitMatch: {
		url: 'https:
	},
	'Strict Equality Comparison': {
		url: 'https:
	},
	StringCreate: {
		url: 'https:
	},
	StringGetIndexProperty: {
		url: 'https:
	},
	SymbolDescriptiveString: {
		url: 'https:
	},
	TestIntegrityLevel: {
		url: 'https:
	},
	thisBooleanValue: {
		url: 'https:
	},
	thisNumberValue: {
		url: 'https:
	},
	thisStringValue: {
		url: 'https:
	},
	thisTimeValue: {
		url: 'https:
	},
	TimeClip: {
		url: 'https:
	},
	TimeFromYear: {
		url: 'https:
	},
	TimeWithinDay: {
		url: 'https:
	},
	ToBoolean: {
		url: 'https:
	},
	ToDateString: {
		url: 'https:
	},
	ToInt16: {
		url: 'https:
	},
	ToInt32: {
		url: 'https:
	},
	ToInt8: {
		url: 'https:
	},
	ToInteger: {
		url: 'https:
	},
	ToLength: {
		url: 'https:
	},
	ToNumber: {
		url: 'https:
	},
	ToObject: {
		url: 'https:
	},
	ToPrimitive: {
		url: 'https:
	},
	ToPropertyDescriptor: {
		url: 'https:
	},
	ToPropertyKey: {
		url: 'https:
	},
	ToString: {
		url: 'https:
	},
	ToUint16: {
		url: 'https:
	},
	ToUint32: {
		url: 'https:
	},
	ToUint8: {
		url: 'https:
	},
	ToUint8Clamp: {
		url: 'https:
	},
	TriggerPromiseReactions: {
		url: 'https:
	},
	Type: {
		url: 'https:
	},
	TypedArrayFrom: {
		url: 'https:
	},
	UpdateEmpty: {
		url: 'https:
	},
	UTC: {
		url: 'https:
	},
	ValidateAndApplyPropertyDescriptor: {
		url: 'https:
	},
	ValidateTypedArray: {
		url: 'https:
	},
	WeekDay: {
		url: 'https:
	},
	YearFromTime: {
		url: 'https:
	}
};
