// es-abstract/operations/2019.js
'use strict';

module.exports = {
	abs: {
		url: 'https:
	},
	'Abstract Equality Comparison': {
		url: 'https:
	},
	'Abstract Relational Comparison': {
		url: 'https:
	},
	AddEntriesFromIterable: {
		url: 'https:
	},
	AddRestrictedFunctionProperties: {
		url: 'https:
	},
	AddWaiter: {
		url: 'https:
	},
	AdvanceStringIndex: {
		url: 'https:
	},
	'agent-order': {
		url: 'https:
	},
	AgentCanSuspend: {
		url: 'https:
	},
	AgentSignifier: {
		url: 'https:
	},
	AllocateArrayBuffer: {
		url: 'https:
	},
	AllocateSharedArrayBuffer: {
		url: 'https:
	},
	AllocateTypedArray: {
		url: 'https:
	},
	AllocateTypedArrayBuffer: {
		url: 'https:
	},
	ArrayCreate: {
		url: 'https:
	},
	ArraySetLength: {
		url: 'https:
	},
	ArraySpeciesCreate: {
		url: 'https:
	},
	AsyncFromSyncIteratorContinuation: {
		url: 'https:
	},
	AsyncFunctionCreate: {
		url: 'https:
	},
	AsyncFunctionStart: {
		url: 'https:
	},
	AsyncGeneratorEnqueue: {
		url: 'https:
	},
	AsyncGeneratorFunctionCreate: {
		url: 'https:
	},
	AsyncGeneratorReject: {
		url: 'https:
	},
	AsyncGeneratorResolve: {
		url: 'https:
	},
	AsyncGeneratorResumeNext: {
		url: 'https:
	},
	AsyncGeneratorStart: {
		url: 'https:
	},
	AsyncGeneratorYield: {
		url: 'https:
	},
	AsyncIteratorClose: {
		url: 'https:
	},
	AtomicLoad: {
		url: 'https:
	},
	AtomicReadModifyWrite: {
		url: 'https:
	},
	Await: {
		url: 'https:
	},
	BackreferenceMatcher: {
		url: 'https:
	},
	BlockDeclarationInstantiation: {
		url: 'https:
	},
	BoundFunctionCreate: {
		url: 'https:
	},
	Call: {
		url: 'https:
	},
	Canonicalize: {
		url: 'https:
	},
	CanonicalNumericIndexString: {
		url: 'https:
	},
	CaseClauseIsSelected: {
		url: 'https:
	},
	CharacterRange: {
		url: 'https:
	},
	CharacterRangeOrUnion: {
		url: 'https:
	},
	CharacterSetMatcher: {
		url: 'https:
	},
	CloneArrayBuffer: {
		url: 'https:
	},
	CompletePropertyDescriptor: {
		url: 'https:
	},
	Completion: {
		url: 'https:
	},
	CompletionRecord: {
		url: 'https:
	},
	ComposeWriteEventBytes: {
		url: 'https:
	},
	Construct: {
		url: 'https:
	},
	CopyDataBlockBytes: {
		url: 'https:
	},
	CopyDataProperties: {
		url: 'https:
	},
	CreateArrayFromList: {
		url: 'https:
	},
	CreateArrayIterator: {
		url: 'https:
	},
	CreateAsyncFromSyncIterator: {
		url: 'https:
	},
	CreateBuiltinFunction: {
		url: 'https:
	},
	CreateByteDataBlock: {
		url: 'https:
	},
	CreateDataProperty: {
		url: 'https:
	},
	CreateDataPropertyOrThrow: {
		url: 'https:
	},
	CreateDynamicFunction: {
		url: 'https:
	},
	CreateHTML: {
		url: 'https:
	},
	CreateIntrinsics: {
		url: 'https:
	},
	CreateIterResultObject: {
		url: 'https:
	},
	CreateListFromArrayLike: {
		url: 'https:
	},
	CreateListIteratorRecord: {
		url: 'https:
	},
	CreateMapIterator: {
		url: 'https:
	},
	CreateMappedArgumentsObject: {
		url: 'https:
	},
	CreateMethodProperty: {
		url: 'https:
	},
	CreatePerIterationEnvironment: {
		url: 'https:
	},
	CreateRealm: {
		url: 'https:
	},
	CreateResolvingFunctions: {
		url: 'https:
	},
	CreateSetIterator: {
		url: 'https:
	},
	CreateSharedByteDataBlock: {
		url: 'https:
	},
	CreateStringIterator: {
		url: 'https:
	},
	CreateUnmappedArgumentsObject: {
		url: 'https:
	},
	DateFromTime: {
		url: 'https:
	},
	DateString: {
		url: 'https:
	},
	Day: {
		url: 'https:
	},
	DayFromYear: {
		url: 'https:
	},
	DaysInYear: {
		url: 'https:
	},
	DayWithinYear: {
		url: 'https:
	},
	Decode: {
		url: 'https:
	},
	DefinePropertyOrThrow: {
		url: 'https:
	},
	DeletePropertyOrThrow: {
		url: 'https:
	},
	DetachArrayBuffer: {
		url: 'https:
	},
	Encode: {
		url: 'https:
	},
	EnqueueJob: {
		url: 'https:
	},
	EnterCriticalSection: {
		url: 'https:
	},
	EnumerableOwnPropertyNames: {
		url: 'https:
	},
	EnumerateObjectProperties: {
		url: 'https:
	},
	EscapeRegExpPattern: {
		url: 'https:
	},
	EvalDeclarationInstantiation: {
		url: 'https:
	},
	EvaluateCall: {
		url: 'https:
	},
	EvaluateNew: {
		url: 'https:
	},
	EventSet: {
		url: 'https:
	},
	ExecuteModule: {
		url: 'https:
	},
	FlattenIntoArray: {
		url: 'https:
	},
	floor: {
		url: 'https:
	},
	ForBodyEvaluation: {
		url: 'https:
	},
	'ForIn/OfBodyEvaluation': {
		url: 'https:
	},
	'ForIn/OfHeadEvaluation': {
		url: 'https:
	},
	FromPropertyDescriptor: {
		url: 'https:
	},
	FulfillPromise: {
		url: 'https:
	},
	FunctionAllocate: {
		url: 'https:
	},
	FunctionCreate: {
		url: 'https:
	},
	FunctionDeclarationInstantiation: {
		url: 'https:
	},
	FunctionInitialize: {
		url: 'https:
	},
	GeneratorFunctionCreate: {
		url: 'https:
	},
	GeneratorResume: {
		url: 'https:
	},
	GeneratorResumeAbrupt: {
		url: 'https:
	},
	GeneratorStart: {
		url: 'https:
	},
	GeneratorValidate: {
		url: 'https:
	},
	GeneratorYield: {
		url: 'https:
	},
	Get: {
		url: 'https:
	},
	GetActiveScriptOrModule: {
		url: 'https:
	},
	GetBase: {
		url: 'https:
	},
	GetFunctionRealm: {
		url: 'https:
	},
	GetGeneratorKind: {
		url: 'https:
	},
	GetGlobalObject: {
		url: 'https:
	},
	GetIdentifierReference: {
		url: 'https:
	},
	GetIterator: {
		url: 'https:
	},
	GetMethod: {
		url: 'https:
	},
	GetModifySetValueInBuffer: {
		url: 'https:
	},
	GetModuleNamespace: {
		url: 'https:
	},
	GetNewTarget: {
		url: 'https:
	},
	GetOwnPropertyKeys: {
		url: 'https:
	},
	GetPrototypeFromConstructor: {
		url: 'https:
	},
	GetReferencedName: {
		url: 'https:
	},
	GetSubstitution: {
		url: 'https:
	},
	GetSuperConstructor: {
		url: 'https:
	},
	GetTemplateObject: {
		url: 'https:
	},
	GetThisEnvironment: {
		url: 'https:
	},
	GetThisValue: {
		url: 'https:
	},
	GetV: {
		url: 'https:
	},
	GetValue: {
		url: 'https:
	},
	GetValueFromBuffer: {
		url: 'https:
	},
	GetViewValue: {
		url: 'https:
	},
	GetWaiterList: {
		url: 'https:
	},
	GlobalDeclarationInstantiation: {
		url: 'https:
	},
	'happens-before': {
		url: 'https:
	},
	HasOwnProperty: {
		url: 'https:
	},
	HasPrimitiveBase: {
		url: 'https:
	},
	HasProperty: {
		url: 'https:
	},
	'host-synchronizes-with': {
		url: 'https:
	},
	HostEventSet: {
		url: 'https:
	},
	HourFromTime: {
		url: 'https:
	},
	IfAbruptRejectPromise: {
		url: 'https:
	},
	ImportedLocalNames: {
		url: 'https:
	},
	InitializeBoundName: {
		url: 'https:
	},
	InitializeEnvironment: {
		url: 'https:
	},
	InitializeHostDefinedRealm: {
		url: 'https:
	},
	InitializeReferencedBinding: {
		url: 'https:
	},
	InLeapYear: {
		url: 'https:
	},
	InnerModuleEvaluation: {
		url: 'https:
	},
	InnerModuleInstantiation: {
		url: 'https:
	},
	InstanceofOperator: {
		url: 'https:
	},
	IntegerIndexedElementGet: {
		url: 'https:
	},
	IntegerIndexedElementSet: {
		url: 'https:
	},
	IntegerIndexedObjectCreate: {
		url: 'https:
	},
	InternalizeJSONProperty: {
		url: 'https:
	},
	Invoke: {
		url: 'https:
	},
	IsAccessorDescriptor: {
		url: 'https:
	},
	IsAnonymousFunctionDefinition: {
		url: 'https:
	},
	IsArray: {
		url: 'https:
	},
	IsCallable: {
		url: 'https:
	},
	IsCompatiblePropertyDescriptor: {
		url: 'https:
	},
	IsConcatSpreadable: {
		url: 'https:
	},
	IsConstructor: {
		url: 'https:
	},
	IsDataDescriptor: {
		url: 'https:
	},
	IsDetachedBuffer: {
		url: 'https:
	},
	IsExtensible: {
		url: 'https:
	},
	IsGenericDescriptor: {
		url: 'https:
	},
	IsInTailPosition: {
		url: 'https:
	},
	IsInteger: {
		url: 'https:
	},
	IsLabelledFunction: {
		url: 'https:
	},
	IsPromise: {
		url: 'https:
	},
	IsPropertyKey: {
		url: 'https:
	},
	IsPropertyReference: {
		url: 'https:
	},
	IsRegExp: {
		url: 'https:
	},
	IsSharedArrayBuffer: {
		url: 'https:
	},
	IsStrictReference: {
		url: 'https:
	},
	IsStringPrefix: {
		url: 'https:
	},
	IsSuperReference: {
		url: 'https:
	},
	IsUnresolvableReference: {
		url: 'https:
	},
	IsWordChar: {
		url: 'https:
	},
	IterableToList: {
		url: 'https:
	},
	IteratorClose: {
		url: 'https:
	},
	IteratorComplete: {
		url: 'https:
	},
	IteratorNext: {
		url: 'https:
	},
	IteratorStep: {
		url: 'https:
	},
	IteratorValue: {
		url: 'https:
	},
	LeaveCriticalSection: {
		url: 'https:
	},
	LocalTime: {
		url: 'https:
	},
	LoopContinues: {
		url: 'https:
	},
	MakeArgGetter: {
		url: 'https:
	},
	MakeArgSetter: {
		url: 'https:
	},
	MakeClassConstructor: {
		url: 'https:
	},
	MakeConstructor: {
		url: 'https:
	},
	MakeDate: {
		url: 'https:
	},
	MakeDay: {
		url: 'https:
	},
	MakeMethod: {
		url: 'https:
	},
	MakeSuperPropertyReference: {
		url: 'https:
	},
	MakeTime: {
		url: 'https:
	},
	max: {
		url: 'https:
	},
	'memory-order': {
		url: 'https:
	},
	min: {
		url: 'https:
	},
	MinFromTime: {
		url: 'https:
	},
	ModuleNamespaceCreate: {
		url: 'https:
	},
	modulo: {
		url: 'https:
	},
	MonthFromTime: {
		url: 'https:
	},
	msFromTime: {
		url: 'https:
	},
	NewDeclarativeEnvironment: {
		url: 'https:
	},
	NewFunctionEnvironment: {
		url: 'https:
	},
	NewGlobalEnvironment: {
		url: 'https:
	},
	NewModuleEnvironment: {
		url: 'https:
	},
	NewObjectEnvironment: {
		url: 'https:
	},
	NewPromiseCapability: {
		url: 'https:
	},
	NormalCompletion: {
		url: 'https:
	},
	NotifyWaiter: {
		url: 'https:
	},
	NumberToRawBytes: {
		url: 'https:
	},
	NumberToString: {
		url: 'https:
	},
	ObjectCreate: {
		url: 'https:
	},
	ObjectDefineProperties: {
		url: 'https:
	},
	OrdinaryCallBindThis: {
		url: 'https:
	},
	OrdinaryCallEvaluateBody: {
		url: 'https:
	},
	OrdinaryCreateFromConstructor: {
		url: 'https:
	},
	OrdinaryDefineOwnProperty: {
		url: 'https:
	},
	OrdinaryDelete: {
		url: 'https:
	},
	OrdinaryGet: {
		url: 'https:
	},
	OrdinaryGetOwnProperty: {
		url: 'https:
	},
	OrdinaryGetPrototypeOf: {
		url: 'https:
	},
	OrdinaryHasInstance: {
		url: 'https:
	},
	OrdinaryHasProperty: {
		url: 'https:
	},
	OrdinaryIsExtensible: {
		url: 'https:
	},
	OrdinaryOwnPropertyKeys: {
		url: 'https:
	},
	OrdinaryPreventExtensions: {
		url: 'https:
	},
	OrdinarySet: {
		url: 'https:
	},
	OrdinarySetPrototypeOf: {
		url: 'https:
	},
	OrdinarySetWithOwnDescriptor: {
		url: 'https:
	},
	OrdinaryToPrimitive: {
		url: 'https:
	},
	ParseModule: {
		url: 'https:
	},
	ParseScript: {
		url: 'https:
	},
	PerformEval: {
		url: 'https:
	},
	PerformPromiseAll: {
		url: 'https:
	},
	PerformPromiseRace: {
		url: 'https:
	},
	PerformPromiseThen: {
		url: 'https:
	},
	PrepareForOrdinaryCall: {
		url: 'https:
	},
	PrepareForTailCall: {
		url: 'https:
	},
	PromiseReactionJob: {
		url: 'https:
	},
	PromiseResolve: {
		url: 'https:
	},
	PromiseResolveThenableJob: {
		url: 'https:
	},
	ProxyCreate: {
		url: 'https:
	},
	PutValue: {
		url: 'https:
	},
	QuoteJSONString: {
		url: 'https:
	},
	RawBytesToNumber: {
		url: 'https:
	},
	'reads-bytes-from': {
		url: 'https:
	},
	'reads-from': {
		url: 'https:
	},
	RegExpAlloc: {
		url: 'https:
	},
	RegExpBuiltinExec: {
		url: 'https:
	},
	RegExpCreate: {
		url: 'https:
	},
	RegExpExec: {
		url: 'https:
	},
	RegExpInitialize: {
		url: 'https:
	},
	RejectPromise: {
		url: 'https:
	},
	RemoveWaiter: {
		url: 'https:
	},
	RemoveWaiters: {
		url: 'https:
	},
	RepeatMatcher: {
		url: 'https:
	},
	RequireObjectCoercible: {
		url: 'https:
	},
	ResolveBinding: {
		url: 'https:
	},
	ResolveThisBinding: {
		url: 'https:
	},
	ReturnIfAbrupt: {
		url: 'https:
	},
	RunJobs: {
		url: 'https:
	},
	SameValue: {
		url: 'https:
	},
	SameValueNonNumber: {
		url: 'https:
	},
	SameValueZero: {
		url: 'https:
	},
	ScriptEvaluation: {
		url: 'https:
	},
	ScriptEvaluationJob: {
		url: 'https:
	},
	SecFromTime: {
		url: 'https:
	},
	SerializeJSONArray: {
		url: 'https:
	},
	SerializeJSONObject: {
		url: 'https:
	},
	SerializeJSONProperty: {
		url: 'https:
	},
	Set: {
		url: 'https:
	},
	SetDefaultGlobalBindings: {
		url: 'https:
	},
	SetFunctionLength: {
		url: 'https:
	},
	SetFunctionName: {
		url: 'https:
	},
	SetImmutablePrototype: {
		url: 'https:
	},
	SetIntegrityLevel: {
		url: 'https:
	},
	SetRealmGlobalObject: {
		url: 'https:
	},
	SetValueInBuffer: {
		url: 'https:
	},
	SetViewValue: {
		url: 'https:
	},
	SharedDataBlockEventSet: {
		url: 'https:
	},
	SortCompare: {
		url: 'https:
	},
	SpeciesConstructor: {
		url: 'https:
	},
	SplitMatch: {
		url: 'https:
	},
	'Strict Equality Comparison': {
		url: 'https:
	},
	StringCreate: {
		url: 'https:
	},
	StringGetOwnProperty: {
		url: 'https:
	},
	Suspend: {
		url: 'https:
	},
	SymbolDescriptiveString: {
		url: 'https:
	},
	SynchronizeEventSet: {
		url: 'https:
	},
	'synchronizes-with': {
		url: 'https:
	},
	TestIntegrityLevel: {
		url: 'https:
	},
	thisBooleanValue: {
		url: 'https:
	},
	thisNumberValue: {
		url: 'https:
	},
	thisStringValue: {
		url: 'https:
	},
	thisSymbolValue: {
		url: 'https:
	},
	thisTimeValue: {
		url: 'https:
	},
	ThrowCompletion: {
		url: 'https:
	},
	TimeClip: {
		url: 'https:
	},
	TimeFromYear: {
		url: 'https:
	},
	TimeString: {
		url: 'https:
	},
	TimeWithinDay: {
		url: 'https:
	},
	TimeZoneString: {
		url: 'https:
	},
	ToBoolean: {
		url: 'https:
	},
	ToDateString: {
		url: 'https:
	},
	ToIndex: {
		url: 'https:
	},
	ToInt16: {
		url: 'https:
	},
	ToInt32: {
		url: 'https:
	},
	ToInt8: {
		url: 'https:
	},
	ToInteger: {
		url: 'https:
	},
	ToLength: {
		url: 'https:
	},
	ToNumber: {
		url: 'https:
	},
	ToObject: {
		url: 'https:
	},
	TopLevelModuleEvaluationJob: {
		url: 'https:
	},
	ToPrimitive: {
		url: 'https:
	},
	ToPropertyDescriptor: {
		url: 'https:
	},
	ToPropertyKey: {
		url: 'https:
	},
	ToString: {
		url: 'https:
	},
	ToUint16: {
		url: 'https:
	},
	ToUint32: {
		url: 'https:
	},
	ToUint8: {
		url: 'https:
	},
	ToUint8Clamp: {
		url: 'https:
	},
	TriggerPromiseReactions: {
		url: 'https:
	},
	TrimString: {
		url: 'https:
	},
	Type: {
		url: 'https:
	},
	TypedArrayCreate: {
		url: 'https:
	},
	TypedArraySpeciesCreate: {
		url: 'https:
	},
	UnicodeEscape: {
		url: 'https:
	},
	UnicodeMatchProperty: {
		url: 'https:
	},
	UnicodeMatchPropertyValue: {
		url: 'https:
	},
	UpdateEmpty: {
		url: 'https:
	},
	UTC: {
		url: 'https:
	},
	UTF16Decode: {
		url: 'https:
	},
	UTF16Encoding: {
		url: 'https:
	},
	ValidateAndApplyPropertyDescriptor: {
		url: 'https:
	},
	ValidateAtomicAccess: {
		url: 'https:
	},
	ValidateSharedIntegerTypedArray: {
		url: 'https:
	},
	ValidateTypedArray: {
		url: 'https:
	},
	ValueOfReadEvent: {
		url: 'https:
	},
	WeekDay: {
		url: 'https:
	},
	WordCharacters: {
		url: 'https:
	},
	YearFromTime: {
		url: 'https:
	}
};
