// es-abstract/operations/2022.js
'use strict';

module.exports = {
	abs: {
		url: 'https:
	},
	AddEntriesFromIterable: {
		url: 'https:
	},
	AddRestrictedFunctionProperties: {
		url: 'https:
	},
	AddToKeptObjects: {
		url: 'https:
	},
	AddWaiter: {
		url: 'https:
	},
	AdvanceStringIndex: {
		url: 'https:
	},
	'agent-order': {
		url: 'https:
	},
	AgentCanSuspend: {
		url: 'https:
	},
	AgentSignifier: {
		url: 'https:
	},
	AllocateArrayBuffer: {
		url: 'https:
	},
	AllocateSharedArrayBuffer: {
		url: 'https:
	},
	AllocateTypedArray: {
		url: 'https:
	},
	AllocateTypedArrayBuffer: {
		url: 'https:
	},
	ApplyStringOrNumericBinaryOperator: {
		url: 'https:
	},
	ArrayCreate: {
		url: 'https:
	},
	ArraySetLength: {
		url: 'https:
	},
	ArraySpeciesCreate: {
		url: 'https:
	},
	AsyncBlockStart: {
		url: 'https:
	},
	AsyncFromSyncIteratorContinuation: {
		url: 'https:
	},
	AsyncFunctionStart: {
		url: 'https:
	},
	AsyncGeneratorAwaitReturn: {
		url: 'https:
	},
	AsyncGeneratorCompleteStep: {
		url: 'https:
	},
	AsyncGeneratorDrainQueue: {
		url: 'https:
	},
	AsyncGeneratorEnqueue: {
		url: 'https:
	},
	AsyncGeneratorResume: {
		url: 'https:
	},
	AsyncGeneratorStart: {
		url: 'https:
	},
	AsyncGeneratorUnwrapYieldResumption: {
		url: 'https:
	},
	AsyncGeneratorValidate: {
		url: 'https:
	},
	AsyncGeneratorYield: {
		url: 'https:
	},
	AsyncIteratorClose: {
		url: 'https:
	},
	AsyncModuleExecutionFulfilled: {
		url: 'https:
	},
	AsyncModuleExecutionRejected: {
		url: 'https:
	},
	AtomicReadModifyWrite: {
		url: 'https:
	},
	Await: {
		url: 'https:
	},
	BackreferenceMatcher: {
		url: 'https:
	},
	'BigInt::add': {
		url: 'https:
	},
	'BigInt::bitwiseAND': {
		url: 'https:
	},
	'BigInt::bitwiseNOT': {
		url: 'https:
	},
	'BigInt::bitwiseOR': {
		url: 'https:
	},
	'BigInt::bitwiseXOR': {
		url: 'https:
	},
	'BigInt::divide': {
		url: 'https:
	},
	'BigInt::equal': {
		url: 'https:
	},
	'BigInt::exponentiate': {
		url: 'https:
	},
	'BigInt::leftShift': {
		url: 'https:
	},
	'BigInt::lessThan': {
		url: 'https:
	},
	'BigInt::multiply': {
		url: 'https:
	},
	'BigInt::remainder': {
		url: 'https:
	},
	'BigInt::sameValue': {
		url: 'https:
	},
	'BigInt::sameValueZero': {
		url: 'https:
	},
	'BigInt::signedRightShift': {
		url: 'https:
	},
	'BigInt::subtract': {
		url: 'https:
	},
	'BigInt::toString': {
		url: 'https:
	},
	'BigInt::unaryMinus': {
		url: 'https:
	},
	'BigInt::unsignedRightShift': {
		url: 'https:
	},
	BigIntBitwiseOp: {
		url: 'https:
	},
	BinaryAnd: {
		url: 'https:
	},
	BinaryOr: {
		url: 'https:
	},
	BinaryXor: {
		url: 'https:
	},
	BlockDeclarationInstantiation: {
		url: 'https:
	},
	BoundFunctionCreate: {
		url: 'https:
	},
	ByteListBitwiseOp: {
		url: 'https:
	},
	ByteListEqual: {
		url: 'https:
	},
	Call: {
		url: 'https:
	},
	Canonicalize: {
		url: 'https:
	},
	CanonicalNumericIndexString: {
		url: 'https:
	},
	CaseClauseIsSelected: {
		url: 'https:
	},
	CharacterRange: {
		url: 'https:
	},
	CharacterRangeOrUnion: {
		url: 'https:
	},
	CharacterSetMatcher: {
		url: 'https:
	},
	clamp: {
		url: 'https:
	},
	CleanupFinalizationRegistry: {
		url: 'https:
	},
	ClearKeptObjects: {
		url: 'https:
	},
	CloneArrayBuffer: {
		url: 'https:
	},
	CodePointAt: {
		url: 'https:
	},
	CodePointsToString: {
		url: 'https:
	},
	CompletePropertyDescriptor: {
		url: 'https:
	},
	Completion: {
		url: 'https:
	},
	CompletionRecord: {
		url: 'https:
	},
	ComposeWriteEventBytes: {
		url: 'https:
	},
	Construct: {
		url: 'https:
	},
	CopyDataBlockBytes: {
		url: 'https:
	},
	CopyDataProperties: {
		url: 'https:
	},
	CreateArrayFromList: {
		url: 'https:
	},
	CreateArrayIterator: {
		url: 'https:
	},
	CreateAsyncFromSyncIterator: {
		url: 'https:
	},
	CreateAsyncIteratorFromClosure: {
		url: 'https:
	},
	CreateBuiltinFunction: {
		url: 'https:
	},
	CreateByteDataBlock: {
		url: 'https:
	},
	CreateDataProperty: {
		url: 'https:
	},
	CreateDataPropertyOrThrow: {
		url: 'https:
	},
	CreateDynamicFunction: {
		url: 'https:
	},
	CreateForInIterator: {
		url: 'https:
	},
	CreateHTML: {
		url: 'https:
	},
	CreateIntrinsics: {
		url: 'https:
	},
	CreateIteratorFromClosure: {
		url: 'https:
	},
	CreateIterResultObject: {
		url: 'https:
	},
	CreateListFromArrayLike: {
		url: 'https:
	},
	CreateListIteratorRecord: {
		url: 'https:
	},
	CreateMapIterator: {
		url: 'https:
	},
	CreateMappedArgumentsObject: {
		url: 'https:
	},
	CreateMethodProperty: {
		url: 'https:
	},
	CreateNonEnumerableDataPropertyOrThrow: {
		url: 'https:
	},
	CreatePerIterationEnvironment: {
		url: 'https:
	},
	CreateRealm: {
		url: 'https:
	},
	CreateRegExpStringIterator: {
		url: 'https:
	},
	CreateResolvingFunctions: {
		url: 'https:
	},
	CreateSetIterator: {
		url: 'https:
	},
	CreateSharedByteDataBlock: {
		url: 'https:
	},
	CreateUnmappedArgumentsObject: {
		url: 'https:
	},
	DateFromTime: {
		url: 'https:
	},
	DateString: {
		url: 'https:
	},
	Day: {
		url: 'https:
	},
	DayFromYear: {
		url: 'https:
	},
	DaysInYear: {
		url: 'https:
	},
	DayWithinYear: {
		url: 'https:
	},
	Decode: {
		url: 'https:
	},
	DefineField: {
		url: 'https:
	},
	DefineMethodProperty: {
		url: 'https:
	},
	DefinePropertyOrThrow: {
		url: 'https:
	},
	DeletePropertyOrThrow: {
		url: 'https:
	},
	DetachArrayBuffer: {
		url: 'https:
	},
	Encode: {
		url: 'https:
	},
	EnterCriticalSection: {
		url: 'https:
	},
	EnumerableOwnPropertyNames: {
		url: 'https:
	},
	EnumerateObjectProperties: {
		url: 'https:
	},
	EscapeRegExpPattern: {
		url: 'https:
	},
	EvalDeclarationInstantiation: {
		url: 'https:
	},
	EvaluateCall: {
		url: 'https:
	},
	EvaluateNew: {
		url: 'https:
	},
	EvaluatePropertyAccessWithExpressionKey: {
		url: 'https:
	},
	EvaluatePropertyAccessWithIdentifierKey: {
		url: 'https:
	},
	EvaluateStringOrNumericBinaryExpression: {
		url: 'https:
	},
	EventSet: {
		url: 'https:
	},
	ExecuteAsyncModule: {
		url: 'https:
	},
	FinishDynamicImport: {
		url: 'https:
	},
	FlattenIntoArray: {
		url: 'https:
	},
	floor: {
		url: 'https:
	},
	ForBodyEvaluation: {
		url: 'https:
	},
	'ForIn/OfBodyEvaluation': {
		url: 'https:
	},
	'ForIn/OfHeadEvaluation': {
		url: 'https:
	},
	FromPropertyDescriptor: {
		url: 'https:
	},
	FulfillPromise: {
		url: 'https:
	},
	FunctionDeclarationInstantiation: {
		url: 'https:
	},
	GatherAvailableAncestors: {
		url: 'https:
	},
	GeneratorResume: {
		url: 'https:
	},
	GeneratorResumeAbrupt: {
		url: 'https:
	},
	GeneratorStart: {
		url: 'https:
	},
	GeneratorValidate: {
		url: 'https:
	},
	GeneratorYield: {
		url: 'https:
	},
	Get: {
		url: 'https:
	},
	GetActiveScriptOrModule: {
		url: 'https:
	},
	GetFunctionRealm: {
		url: 'https:
	},
	GetGeneratorKind: {
		url: 'https:
	},
	GetGlobalObject: {
		url: 'https:
	},
	GetIdentifierReference: {
		url: 'https:
	},
	GetIterator: {
		url: 'https:
	},
	GetMatchIndexPair: {
		url: 'https:
	},
	GetMatchString: {
		url: 'https:
	},
	GetMethod: {
		url: 'https:
	},
	GetModifySetValueInBuffer: {
		url: 'https:
	},
	GetModuleNamespace: {
		url: 'https:
	},
	GetNewTarget: {
		url: 'https:
	},
	GetOwnPropertyKeys: {
		url: 'https:
	},
	GetPromiseResolve: {
		url: 'https:
	},
	GetPrototypeFromConstructor: {
		url: 'https:
	},
	GetStringIndex: {
		url: 'https:
	},
	GetSubstitution: {
		url: 'https:
	},
	GetSuperConstructor: {
		url: 'https:
	},
	GetTemplateObject: {
		url: 'https:
	},
	GetThisEnvironment: {
		url: 'https:
	},
	GetThisValue: {
		url: 'https:
	},
	GetV: {
		url: 'https:
	},
	GetValue: {
		url: 'https:
	},
	GetValueFromBuffer: {
		url: 'https:
	},
	GetViewValue: {
		url: 'https:
	},
	GetWaiterList: {
		url: 'https:
	},
	GlobalDeclarationInstantiation: {
		url: 'https:
	},
	'happens-before': {
		url: 'https:
	},
	HasOwnProperty: {
		url: 'https:
	},
	HasProperty: {
		url: 'https:
	},
	'host-synchronizes-with': {
		url: 'https:
	},
	HostEventSet: {
		url: 'https:
	},
	HourFromTime: {
		url: 'https:
	},
	IfAbruptCloseIterator: {
		url: 'https:
	},
	IfAbruptRejectPromise: {
		url: 'https:
	},
	ImportedLocalNames: {
		url: 'https:
	},
	InitializeBoundName: {
		url: 'https:
	},
	InitializeHostDefinedRealm: {
		url: 'https:
	},
	InitializeInstanceElements: {
		url: 'https:
	},
	InitializeReferencedBinding: {
		url: 'https:
	},
	InitializeTypedArrayFromArrayBuffer: {
		url: 'https:
	},
	InitializeTypedArrayFromArrayLike: {
		url: 'https:
	},
	InitializeTypedArrayFromList: {
		url: 'https:
	},
	InitializeTypedArrayFromTypedArray: {
		url: 'https:
	},
	InLeapYear: {
		url: 'https:
	},
	InnerModuleEvaluation: {
		url: 'https:
	},
	InnerModuleLinking: {
		url: 'https:
	},
	InstallErrorCause: {
		url: 'https:
	},
	InstanceofOperator: {
		url: 'https:
	},
	IntegerIndexedElementGet: {
		url: 'https:
	},
	IntegerIndexedElementSet: {
		url: 'https:
	},
	IntegerIndexedObjectCreate: {
		url: 'https:
	},
	InternalizeJSONProperty: {
		url: 'https:
	},
	Invoke: {
		url: 'https:
	},
	IsAccessorDescriptor: {
		url: 'https:
	},
	IsAnonymousFunctionDefinition: {
		url: 'https:
	},
	IsArray: {
		url: 'https:
	},
	IsBigIntElementType: {
		url: 'https:
	},
	IsCallable: {
		url: 'https:
	},
	IsCompatiblePropertyDescriptor: {
		url: 'https:
	},
	IsConcatSpreadable: {
		url: 'https:
	},
	IsConstructor: {
		url: 'https:
	},
	IsDataDescriptor: {
		url: 'https:
	},
	IsDetachedBuffer: {
		url: 'https:
	},
	IsExtensible: {
		url: 'https:
	},
	IsGenericDescriptor: {
		url: 'https:
	},
	IsInTailPosition: {
		url: 'https:
	},
	IsIntegralNumber: {
		url: 'https:
	},
	IsLabelledFunction: {
		url: 'https:
	},
	IsLessThan: {
		url: 'https:
	},
	IsLooselyEqual: {
		url: 'https:
	},
	IsNoTearConfiguration: {
		url: 'https:
	},
	IsPrivateReference: {
		url: 'https:
	},
	IsPromise: {
		url: 'https:
	},
	IsPropertyKey: {
		url: 'https:
	},
	IsPropertyReference: {
		url: 'https:
	},
	IsRegExp: {
		url: 'https:
	},
	IsSharedArrayBuffer: {
		url: 'https:
	},
	IsStrictlyEqual: {
		url: 'https:
	},
	IsStringPrefix: {
		url: 'https:
	},
	IsStringWellFormedUnicode: {
		url: 'https:
	},
	IsSuperReference: {
		url: 'https:
	},
	IsUnclampedIntegerElementType: {
		url: 'https:
	},
	IsUnresolvableReference: {
		url: 'https:
	},
	IsUnsignedElementType: {
		url: 'https:
	},
	IsValidIntegerIndex: {
		url: 'https:
	},
	IsValidRegularExpressionLiteral: {
		url: 'https:
	},
	IsWordChar: {
		url: 'https:
	},
	IterableToList: {
		url: 'https:
	},
	IteratorClose: {
		url: 'https:
	},
	IteratorComplete: {
		url: 'https:
	},
	IteratorNext: {
		url: 'https:
	},
	IteratorStep: {
		url: 'https:
	},
	IteratorValue: {
		url: 'https:
	},
	LeaveCriticalSection: {
		url: 'https:
	},
	LengthOfArrayLike: {
		url: 'https:
	},
	LocalTime: {
		url: 'https:
	},
	LocalTZA: {
		url: 'https:
	},
	LoopContinues: {
		url: 'https:
	},
	MakeArgGetter: {
		url: 'https:
	},
	MakeArgSetter: {
		url: 'https:
	},
	MakeBasicObject: {
		url: 'https:
	},
	MakeClassConstructor: {
		url: 'https:
	},
	MakeConstructor: {
		url: 'https:
	},
	MakeDate: {
		url: 'https:
	},
	MakeDay: {
		url: 'https:
	},
	MakeMatchIndicesIndexPairArray: {
		url: 'https:
	},
	MakeMethod: {
		url: 'https:
	},
	MakePrivateReference: {
		url: 'https:
	},
	MakeSuperPropertyReference: {
		url: 'https:
	},
	MakeTime: {
		url: 'https:
	},
	max: {
		url: 'https:
	},
	'memory-order': {
		url: 'https:
	},
	min: {
		url: 'https:
	},
	MinFromTime: {
		url: 'https:
	},
	ModuleNamespaceCreate: {
		url: 'https:
	},
	modulo: {
		url: 'https:
	},
	MonthFromTime: {
		url: 'https:
	},
	msFromTime: {
		url: 'https:
	},
	NewDeclarativeEnvironment: {
		url: 'https:
	},
	NewFunctionEnvironment: {
		url: 'https:
	},
	NewGlobalEnvironment: {
		url: 'https:
	},
	NewModuleEnvironment: {
		url: 'https:
	},
	NewObjectEnvironment: {
		url: 'https:
	},
	NewPrivateEnvironment: {
		url: 'https:
	},
	NewPromiseCapability: {
		url: 'https:
	},
	NewPromiseReactionJob: {
		url: 'https:
	},
	NewPromiseResolveThenableJob: {
		url: 'https:
	},
	NormalCompletion: {
		url: 'https:
	},
	NotifyWaiter: {
		url: 'https:
	},
	'Number::add': {
		url: 'https:
	},
	'Number::bitwiseAND': {
		url: 'https:
	},
	'Number::bitwiseNOT': {
		url: 'https:
	},
	'Number::bitwiseOR': {
		url: 'https:
	},
	'Number::bitwiseXOR': {
		url: 'https:
	},
	'Number::divide': {
		url: 'https:
	},
	'Number::equal': {
		url: 'https:
	},
	'Number::exponentiate': {
		url: 'https:
	},
	'Number::leftShift': {
		url: 'https:
	},
	'Number::lessThan': {
		url: 'https:
	},
	'Number::multiply': {
		url: 'https:
	},
	'Number::remainder': {
		url: 'https:
	},
	'Number::sameValue': {
		url: 'https:
	},
	'Number::sameValueZero': {
		url: 'https:
	},
	'Number::signedRightShift': {
		url: 'https:
	},
	'Number::subtract': {
		url: 'https:
	},
	'Number::toString': {
		url: 'https:
	},
	'Number::unaryMinus': {
		url: 'https:
	},
	'Number::unsignedRightShift': {
		url: 'https:
	},
	NumberBitwiseOp: {
		url: 'https:
	},
	NumberToBigInt: {
		url: 'https:
	},
	NumericToRawBytes: {
		url: 'https:
	},
	ObjectDefineProperties: {
		url: 'https:
	},
	OrdinaryCallBindThis: {
		url: 'https:
	},
	OrdinaryCallEvaluateBody: {
		url: 'https:
	},
	OrdinaryCreateFromConstructor: {
		url: 'https:
	},
	OrdinaryDefineOwnProperty: {
		url: 'https:
	},
	OrdinaryDelete: {
		url: 'https:
	},
	OrdinaryFunctionCreate: {
		url: 'https:
	},
	OrdinaryGet: {
		url: 'https:
	},
	OrdinaryGetOwnProperty: {
		url: 'https:
	},
	OrdinaryGetPrototypeOf: {
		url: 'https:
	},
	OrdinaryHasInstance: {
		url: 'https:
	},
	OrdinaryHasProperty: {
		url: 'https:
	},
	OrdinaryIsExtensible: {
		url: 'https:
	},
	OrdinaryObjectCreate: {
		url: 'https:
	},
	OrdinaryOwnPropertyKeys: {
		url: 'https:
	},
	OrdinaryPreventExtensions: {
		url: 'https:
	},
	OrdinarySet: {
		url: 'https:
	},
	OrdinarySetPrototypeOf: {
		url: 'https:
	},
	OrdinarySetWithOwnDescriptor: {
		url: 'https:
	},
	OrdinaryToPrimitive: {
		url: 'https:
	},
	ParseModule: {
		url: 'https:
	},
	ParsePattern: {
		url: 'https:
	},
	ParseScript: {
		url: 'https:
	},
	ParseText: {
		url: 'https:
	},
	PerformEval: {
		url: 'https:
	},
	PerformPromiseAll: {
		url: 'https:
	},
	PerformPromiseAllSettled: {
		url: 'https:
	},
	PerformPromiseAny: {
		url: 'https:
	},
	PerformPromiseRace: {
		url: 'https:
	},
	PerformPromiseThen: {
		url: 'https:
	},
	PrepareForOrdinaryCall: {
		url: 'https:
	},
	PrepareForTailCall: {
		url: 'https:
	},
	PrivateElementFind: {
		url: 'https:
	},
	PrivateFieldAdd: {
		url: 'https:
	},
	PrivateGet: {
		url: 'https:
	},
	PrivateMethodOrAccessorAdd: {
		url: 'https:
	},
	PrivateSet: {
		url: 'https:
	},
	PromiseResolve: {
		url: 'https:
	},
	ProxyCreate: {
		url: 'https:
	},
	PutValue: {
		url: 'https:
	},
	QuoteJSONString: {
		url: 'https:
	},
	RawBytesToNumeric: {
		url: 'https:
	},
	'reads-bytes-from': {
		url: 'https:
	},
	'reads-from': {
		url: 'https:
	},
	RegExpAlloc: {
		url: 'https:
	},
	RegExpBuiltinExec: {
		url: 'https:
	},
	RegExpCreate: {
		url: 'https:
	},
	RegExpExec: {
		url: 'https:
	},
	RegExpHasFlag: {
		url: 'https:
	},
	RegExpInitialize: {
		url: 'https:
	},
	RejectPromise: {
		url: 'https:
	},
	RemoveWaiter: {
		url: 'https:
	},
	RemoveWaiters: {
		url: 'https:
	},
	RepeatMatcher: {
		url: 'https:
	},
	RequireInternalSlot: {
		url: 'https:
	},
	RequireObjectCoercible: {
		url: 'https:
	},
	ResolveBinding: {
		url: 'https:
	},
	ResolvePrivateIdentifier: {
		url: 'https:
	},
	ResolveThisBinding: {
		url: 'https:
	},
	ReturnIfAbrupt: {
		url: 'https:
	},
	RoundMVResult: {
		url: 'https:
	},
	SameValue: {
		url: 'https:
	},
	SameValueNonNumeric: {
		url: 'https:
	},
	SameValueZero: {
		url: 'https:
	},
	ScriptEvaluation: {
		url: 'https:
	},
	SecFromTime: {
		url: 'https:
	},
	SerializeJSONArray: {
		url: 'https:
	},
	SerializeJSONObject: {
		url: 'https:
	},
	SerializeJSONProperty: {
		url: 'https:
	},
	Set: {
		url: 'https:
	},
	SetDefaultGlobalBindings: {
		url: 'https:
	},
	SetFunctionLength: {
		url: 'https:
	},
	SetFunctionName: {
		url: 'https:
	},
	SetImmutablePrototype: {
		url: 'https:
	},
	SetIntegrityLevel: {
		url: 'https:
	},
	SetRealmGlobalObject: {
		url: 'https:
	},
	SetTypedArrayFromArrayLike: {
		url: 'https:
	},
	SetTypedArrayFromTypedArray: {
		url: 'https:
	},
	SetValueInBuffer: {
		url: 'https:
	},
	SetViewValue: {
		url: 'https:
	},
	SharedDataBlockEventSet: {
		url: 'https:
	},
	SortIndexedProperties: {
		url: 'https:
	},
	SpeciesConstructor: {
		url: 'https:
	},
	StringCreate: {
		url: 'https:
	},
	StringGetOwnProperty: {
		url: 'https:
	},
	StringIndexOf: {
		url: 'https:
	},
	StringPad: {
		url: 'https:
	},
	StringToBigInt: {
		url: 'https:
	},
	StringToCodePoints: {
		url: 'https:
	},
	StringToNumber: {
		url: 'https:
	},
	substring: {
		url: 'https:
	},
	SuspendAgent: {
		url: 'https:
	},
	SymbolDescriptiveString: {
		url: 'https:
	},
	'synchronizes-with': {
		url: 'https:
	},
	TestIntegrityLevel: {
		url: 'https:
	},
	thisBigIntValue: {
		url: 'https:
	},
	thisBooleanValue: {
		url: 'https:
	},
	thisNumberValue: {
		url: 'https:
	},
	thisStringValue: {
		url: 'https:
	},
	thisSymbolValue: {
		url: 'https:
	},
	thisTimeValue: {
		url: 'https:
	},
	ThrowCompletion: {
		url: 'https:
	},
	TimeClip: {
		url: 'https:
	},
	TimeFromYear: {
		url: 'https:
	},
	TimeString: {
		url: 'https:
	},
	TimeWithinDay: {
		url: 'https:
	},
	TimeZoneString: {
		url: 'https:
	},
	ToBigInt: {
		url: 'https:
	},
	ToBigInt64: {
		url: 'https:
	},
	ToBigUint64: {
		url: 'https:
	},
	ToBoolean: {
		url: 'https:
	},
	ToDateString: {
		url: 'https:
	},
	ToIndex: {
		url: 'https:
	},
	ToInt16: {
		url: 'https:
	},
	ToInt32: {
		url: 'https:
	},
	ToInt8: {
		url: 'https:
	},
	ToIntegerOrInfinity: {
		url: 'https:
	},
	ToLength: {
		url: 'https:
	},
	ToNumber: {
		url: 'https:
	},
	ToNumeric: {
		url: 'https:
	},
	ToObject: {
		url: 'https:
	},
	ToPrimitive: {
		url: 'https:
	},
	ToPropertyDescriptor: {
		url: 'https:
	},
	ToPropertyKey: {
		url: 'https:
	},
	ToString: {
		url: 'https:
	},
	ToUint16: {
		url: 'https:
	},
	ToUint32: {
		url: 'https:
	},
	ToUint8: {
		url: 'https:
	},
	ToUint8Clamp: {
		url: 'https:
	},
	ToZeroPaddedDecimalString: {
		url: 'https:
	},
	TriggerPromiseReactions: {
		url: 'https:
	},
	TrimString: {
		url: 'https:
	},
	Type: {
		url: 'https:
	},
	TypedArrayCreate: {
		url: 'https:
	},
	TypedArrayElementSize: {
		url: 'https:
	},
	TypedArrayElementType: {
		url: 'https:
	},
	TypedArraySpeciesCreate: {
		url: 'https:
	},
	UnicodeEscape: {
		url: 'https:
	},
	UnicodeMatchProperty: {
		url: 'https:
	},
	UnicodeMatchPropertyValue: {
		url: 'https:
	},
	UpdateEmpty: {
		url: 'https:
	},
	UTC: {
		url: 'https:
	},
	UTF16EncodeCodePoint: {
		url: 'https:
	},
	UTF16SurrogatePairToCodePoint: {
		url: 'https:
	},
	ValidateAndApplyPropertyDescriptor: {
		url: 'https:
	},
	ValidateAtomicAccess: {
		url: 'https:
	},
	ValidateIntegerTypedArray: {
		url: 'https:
	},
	ValidateTypedArray: {
		url: 'https:
	},
	ValueOfReadEvent: {
		url: 'https:
	},
	WeakRefDeref: {
		url: 'https:
	},
	WeekDay: {
		url: 'https:
	},
	WordCharacters: {
		url: 'https:
	},
	YearFromTime: {
		url: 'https:
	},
	Yield: {
		url: 'https:
	}
};
