// es-abstract/operations/es5.js
'use strict';

module.exports = {
	'Abstract Equality Comparison': 'https:
	'Abstract Relational Comparison': 'https:
	'Strict Equality Comparison': 'https:
	abs: 'http:
	Canonicalize: 'https:
	CheckObjectCoercible: 'https:
	DateFromTime: 'https:
	Day: 'https:
	DayFromYear: 'https:
	DaysInYear: 'https:
	DayWithinYear: 'https:
	floor: 'http:
	FromPropertyDescriptor: 'https:
	HourFromTime: 'https:
	InLeapYear: 'https:
	IsAccessorDescriptor: 'https:
	IsCallable: 'https:
	IsDataDescriptor: 'https:
	IsGenericDescriptor: 'https:
	IsPropertyDescriptor: 'https:
	MakeDate: 'https:
	MakeDay: 'https:
	MakeTime: 'https:
	MinFromTime: 'https:
	modulo: 'https:
	MonthFromTime: 'https:
	msFromTime: 'https:
	SameValue: 'https:
	SecFromTime: 'https:
	SplitMatch: 'https:
	TimeClip: 'https:
	TimeFromYear: 'https:
	TimeWithinDay: 'https:
	ToBoolean: 'https:
	ToInt32: 'https:
	ToInteger: 'https:
	ToNumber: 'https:
	ToObject: 'https:
	ToPrimitive: 'https:
	ToPropertyDescriptor: 'https:
	ToString: 'https:
	ToUint16: 'https:
	ToUint32: 'https:
	Type: 'https:
	WeekDay: 'https:
	YearFromTime: 'https:
};
