// istanbul-lib-report/lib/tree.js

"use strict";

var util = require('util');

function Visitor(delegate) {
    this.delegate = delegate;
}

['Start', 'End', 'Summary', 'SummaryEnd', 'Detail' ].forEach(function (k) {
    var f = 'on' + k;
    Visitor.prototype[f] = function (node, state) {
        if (this.delegate[f] && typeof this.delegate[f] === 'function') {
            this.delegate[f].call(this.delegate, node, state);
        }
    };
});

function CompositeVisitor(visitors) {
    if (!Array.isArray(visitors)) {
        visitors = [visitors];
    }
    this.visitors = visitors.map(function (v) {
        if (v instanceof Visitor) {
            return v;
        }
        return new Visitor(v);
    });
}

util.inherits(CompositeVisitor, Visitor);

['Start', 'Summary', 'SummaryEnd', 'Detail', 'End'].forEach(function (k) {
    var f = 'on' + k;
    CompositeVisitor.prototype[f] = function (node, state) {
        this.visitors.forEach(function (v) {
            v[f](node, state);
        });
    };
});

function Node() {
}


Node.prototype.getQualifiedName = function () {
    throw new Error('getQualifiedName must be overridden');
};


Node.prototype.getRelativeName = function () {
    throw new Error('getRelativeName must be overridden');
};


Node.prototype.isRoot = function () {
    return !this.getParent();
};


Node.prototype.getParent = function () {
    throw new Error('getParent must be overridden');
};


Node.prototype.getChildren = function () {
    throw new Error('getChildren must be overridden');
};


Node.prototype.isSummary = function () {
    throw new Error('isSummary must be overridden');
};


Node.prototype.getCoverageSummary = function () {
    throw new Error('getCoverageSummary must be overridden');
};


Node.prototype.getFileCoverage = function () {
    throw new Error('getFileCoverage must be overridden');
};

Node.prototype.visit = function (visitor, state) {

    var that = this,
        visitChildren = function () {
            that.getChildren().forEach(function (child) {
                child.visit(visitor, state);
            });
        };

    if (this.isSummary()) {
        visitor.onSummary(this, state);
    } else {
        visitor.onDetail(this, state);
    }

    visitChildren();

    if (this.isSummary()) {
        visitor.onSummaryEnd(this, state);
    }
};


function Tree() {
}



Tree.prototype.getRoot = function () {
    throw new Error('getRoot must be overridden');
};


Tree.prototype.visit = function (visitor, state) {
    if (!(visitor instanceof Visitor)) {
        visitor = new Visitor(visitor);
    }
    visitor.onStart(this.getRoot(), state);
    this.getRoot().visit(visitor, state);
    visitor.onEnd(this.getRoot(), state);
};

module.exports = {
    Tree: Tree,
    Node: Node,
    Visitor: Visitor,
    CompositeVisitor: CompositeVisitor
};
