// istanbul-lib-report/lib/xml-writer.js

var INDENT = '  ';


function XMLWriter(contentWriter) {
    this.cw = contentWriter;
    this.stack = [];
}

function attrString(attrs) {
    if (!attrs) {
        return '';
    }
    var ret = [];
    Object.keys(attrs).forEach(function (k) {
        var v = attrs[k];
        ret.push(k + '="' + v + '"');
    });
    return ret.length === 0 ? '' : ' ' + ret.join(' ');
}

XMLWriter.prototype.indent = function (str) {
    return this.stack.map(function () { return INDENT; }).join('') + str;
};


XMLWriter.prototype.openTag = function (name, attrs) {
    var str = this.indent('<' + name + attrString(attrs) + '>');
    this.cw.println(str);
    this.stack.push(name);
};


XMLWriter.prototype.closeTag = function (name) {
    if (this.stack.length === 0) {
        throw new Error('Attempt to close tag ' + name + ' when not opened');
    }
    var stashed = this.stack.pop(),
        str = '</' + name + '>';

    if (stashed !== name) {
        throw new Error('Attempt to close tag ' + name + ' when ' + stashed + ' was the one open');
    }
    this.cw.println(this.indent(str));
};

XMLWriter.prototype.inlineTag = function (name, attrs, content) {
    var str = '<' + name + attrString(attrs);
    if (content) {
        str += '>' + content + '</' + name + '>';
    } else {
        str += '/>';
    }
    str = this.indent(str);
    this.cw.println(str);
};

XMLWriter.prototype.closeAll = function () {
    var that = this;
    this.stack.slice().reverse().forEach(function (name) {
        that.closeTag(name);
    });
};

module.exports = XMLWriter;
