// jsdom/lib/jsdom/living/window/History-impl.js
"use strict";
const DOMException = require("domexception");
const { documentBaseURLSerialized, parseURLToResultingURLRecord } = require("../helpers/document-base-url.js");


exports.implementation = class HistoryImpl {
  constructor(args, privateData) {
    this._window = privateData.window;
    this._document = privateData.document;
    this._actAsIfLocationReloadCalled = privateData.actAsIfLocationReloadCalled;
    this._state = null;
  }

  _guardAgainstInactiveDocuments() {
    if (!this._window) {
      throw new DOMException("History object is associated with a document that is not fully active.", "SecurityError");
    }
  }

  get length() {
    this._guardAgainstInactiveDocuments();

    return this._window._sessionHistory.length;
  }

  get state() {
    this._guardAgainstInactiveDocuments();

    return this._state;
  }

  go(delta) {
    this._guardAgainstInactiveDocuments();

    if (delta === 0) {
      
      
      this._actAsIfLocationReloadCalled();
    } else {
      
      this._window._sessionHistory.traverseByDelta(delta);
    }
  }

  back() {
    this.go(-1);
  }

  forward() {
    this.go(+1);
  }

  pushState(data, title, url) {
    this._sharedPushAndReplaceState(data, title, url, "pushState");
  }
  replaceState(data, title, url) {
    this._sharedPushAndReplaceState(data, title, url, "replaceState");
  }

  
  _sharedPushAndReplaceState(data, title, url, methodName) {
    this._guardAgainstInactiveDocuments();

    

    let newURL;
    if (url !== null) {
      
      
      newURL = parseURLToResultingURLRecord(url, this._document);

      if (newURL === null) {
        throw new DOMException(`Could not parse url argument "${url}" to ${methodName} ` +
          `against base URL "${documentBaseURLSerialized(this._document)}".`, "SecurityError");
      }

      if (newURL.scheme !== this._document._URL.scheme ||
          newURL.username !== this._document._URL.username ||
          newURL.password !== this._document._URL.password ||
          newURL.host !== this._document._URL.host ||
          newURL.port !== this._document._URL.port ||
          newURL.cannotBeABaseURL !== this._document._URL.cannotBeABaseURL) {
        throw new DOMException(`${methodName} cannot update history to a URL which ` +
          `differs in components other than in path, query, or fragment.`, "SecurityError");
      }

      
    } else {
      newURL = this._window._sessionHistory.currentEntry.url;
    }

    if (methodName === "pushState") {
      this._window._sessionHistory.removeAllEntriesAfterCurrentEntry();

      this._window._sessionHistory.clearHistoryTraversalTasks();

      const newEntry = {
        document: this._document,
        stateObject: data,
        title,
        url: newURL
      };
      this._window._sessionHistory.addEntryAfterCurrentEntry(newEntry);
      this._window._sessionHistory.updateCurrentEntry(newEntry);
    } else {
      const { currentEntry } = this._window._sessionHistory;
      currentEntry.stateObject = data;
      currentEntry.title = title;
      currentEntry.url = newURL;
    }

    
    
    

    this._document._URL = newURL;

    
    
    this._state = data; 
    this._document._latestEntry = this._window._sessionHistory.currentEntry;
  }
};
