// jsdom/lib/jsdom/named-properties-tracker.js
"use strict";


const IS_NAMED_PROPERTY = Symbol("is named property");
const TRACKER = Symbol("named property tracker");


exports.create = function (object, objectProxy, resolverFunc) {
  if (object[TRACKER]) {
    throw Error("A NamedPropertiesTracker has already been created for this object");
  }

  const tracker = new NamedPropertiesTracker(object, objectProxy, resolverFunc);
  object[TRACKER] = tracker;
  return tracker;
};

exports.get = function (object) {
  if (!object) {
    return null;
  }

  return object[TRACKER] || null;
};

function NamedPropertiesTracker(object, objectProxy, resolverFunc) {
  this.object = object;
  this.objectProxy = objectProxy;
  this.resolverFunc = resolverFunc;
  this.trackedValues = new Map(); 
}

function newPropertyDescriptor(tracker, name) {
  const emptySet = new Set();

  function getValues() {
    return tracker.trackedValues.get(name) || emptySet;
  }

  const descriptor = {
    enumerable: true,
    configurable: true,
    get() {
      return tracker.resolverFunc(tracker.object, name, getValues);
    },
    set(value) {
      Object.defineProperty(tracker.object, name, {
        enumerable: true,
        configurable: true,
        writable: true,
        value
      });
    }
  };

  descriptor.get[IS_NAMED_PROPERTY] = true;
  descriptor.set[IS_NAMED_PROPERTY] = true;
  return descriptor;
}


NamedPropertiesTracker.prototype.track = function (name, value) {
  if (name === undefined || name === null || name === "") {
    return;
  }

  let valueSet = this.trackedValues.get(name);
  if (!valueSet) {
    valueSet = new Set();
    this.trackedValues.set(name, valueSet);
  }

  valueSet.add(value);

  if (name in this.objectProxy) {
    
    return;
  }

  const descriptor = newPropertyDescriptor(this, name);
  Object.defineProperty(this.object, name, descriptor);
};


NamedPropertiesTracker.prototype.untrack = function (name, value) {
  if (name === undefined || name === null || name === "") {
    return;
  }

  const valueSet = this.trackedValues.get(name);
  if (!valueSet) {
    
    return;
  }

  if (!valueSet.delete(value)) {
    
    return;
  }

  if (valueSet.size === 0) {
    this.trackedValues.delete(name);
  }

  if (valueSet.size > 0) {
    
    return;
  }

  

  const descriptor = Object.getOwnPropertyDescriptor(this.object, name);

  if (!descriptor || !descriptor.get || descriptor.get[IS_NAMED_PROPERTY] !== true) {
    
    return;
  }

  
  
  
  delete this.object[name];
};
